compiler/modules/nback/function.bal

```diff
@@ -41,83 +41,113 @@ final RuntimeFunction addToRestArgsFunction = {
     attrs: []
 };
 
-function buildCall(llvm:Builder builder, Scaffold scaffold, bir:CallInsn insn) returns BuildError? {
-    var { symbol: funcSymbol, erasedSignature, signature } = insn.func;
+type FunctionReferenceValue record {|
     llvm:Function func;
-    if funcSymbol is bir:InternalSymbol {
-        func = scaffold.getFunctionDefn(funcSymbol.identifier);
+    t:FunctionSignature signature;
+    t:FunctionSignature erasedSignature;
+|};
+
+type FunctionPointerValue record {|
+    llvm:PointerValue funcValuePtr;
+    llvm:PointerValue funcDescPtr;
+    llvm:PointerValue funcPtr;
+    llvm:PointerValue uniformFuncPtr;
+|};
+
+type FunctionValue FunctionReferenceValue|FunctionPointerValue;
+
+function buildCall(llvm:Builder builder, Scaffold scaffold, bir:CallInsn insn) returns BuildError? {
+    bir:Operand[] args = insn.operands.slice(1);
+    bir:Register result = insn.result;
+    bir:FunctionOperand funcOperand = insn.operands[0];
+    FunctionValue funcVal = check buildFunctionValue(builder, scaffold, funcOperand);
+    if funcVal is FunctionReferenceValue {
+        return buildDirectCall(builder, scaffold, funcVal.func, funcVal.erasedSignature,
+                               funcVal.signature, funcOperand, args, result, funcVal.erasedSignature.returnType);
     }
-    else {
-        func = check buildFunctionDecl(scaffold, funcSymbol, erasedSignature);
+    if insn !is bir:CallIndirectInsn {
+        panic err:impossible("unexpected combination of call instruction and function value");
     }
```

Comment:
We shouldn't have to check this at runtime. We restricted this at type level.

Reply:
Fixed (by splitting into`buildCall` and `buildCallIndirect`)